pyrefly/lib/alt/function.rs

```diff
@@ -884,54 +884,91 @@ impl<'a, Ans: LookupAnswer> AnswersSolver<'a, Ans> {
+                    let call_attr = self.instance_as_dunder_call(&cls).and_then(|call_attr| {
+                        if let Type::BoundMethod(m) = call_attr {
+                            Some(
+                                self.bind_boundmethod(&m, &mut |a, b| self.is_subset_eq(a, b))
+                                    .unwrap_or(m.func.as_type()),
+                            )
+                        } else {
+                            None
+                        }
+                    });
+                    if let Some(mut call_attr) = call_attr {
+                        call_attr.transform_toplevel_func_metadata(|m| {
+                            *m = FuncMetadata {
+                                kind: FunctionKind::CallbackProtocol(Box::new(cls.clone())),
+                                flags: metadata.flags.clone(),
+                            };
+                        });
+                        call_attr
                     } else {
-                        None
+                        cls.to_type()
                     }
-                });
-                if let Some(mut call_attr) = call_attr {
-                    call_attr.transform_toplevel_func_metadata(|m| {
-                        *m = FuncMetadata {
-                            kind: FunctionKind::CallbackProtocol(Box::new(cls.clone())),
-                            flags: metadata.flags.clone(),
-                        };
-                    });
-                    call_attr
-                } else {
-                    cls.to_type()
+                }
+                Type::ClassType(cls) if cls.has_qname("functools", "_Wrapped") => decoratee.clone(),
+                returned_ty => returned_ty,
+            };
+
+        // Re-wrap with *relevant* original Type Parameters.
+        if let Some(tparams) = tparams_opt {
+            if !matches!(inferred_ty, Type::Forall(_)) {
+                // FIX: Wrap this analysis in a block { ... }
+                // This ensures `used_quantifieds` (which borrows `inferred_ty`) is dropped
+                // BEFORE we try to move `inferred_ty` in the match statement below.
+                let relevant_tparams: Vec<TParam> = {
+                    let mut used_quantifieds = SmallSet::new();
+                    inferred_ty.collect_quantifieds(&mut used_quantifieds);
+
+                    tparams
+                        .iter()
+                        .filter(|p| used_quantifieds.contains(&p.quantified))
+                        .cloned()
+                        .collect()
+                };
+
+                if !relevant_tparams.is_empty() {
+                    let new_tparams = Arc::new(TParams::new(relevant_tparams));
+                    return match inferred_ty {
+                        Type::Function(f) => Forallable::Function(*f).forall(new_tparams),
+                        Type::Callable(c) => Forallable::Callable(*c).forall(new_tparams),
+                        _ => inferred_ty,
```

Comment:
I think we're still missing logic here - we actually know for sure that `relevant_tparams` is nonempty, which means we'll definitely leak a `Quantified` if we don't promote it here.

The example I have in mind is still
```
def decorator(f: Callable[[T], T]) -> list[T]): ...
@decorator
def f(x): ...
```
which I think would result in `list[@T]` at this point, we do need to sweep up the unusable quantifieds and turn it into `list[Any]`